microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/PropertiesManagerForJava.java

```diff
@@ -370,6 +371,42 @@ public Hover hover(MicroProfileJavaHoverParams params, IJDTUtils utils, IProgres
 		return hovers.get(0);
 	}
 
+	/**
+	 * Returns the cursor context for the given file and cursor position, or NONE_OF_THE_ABOVE if none of the cursor contexts are applicable.
+	 *
+	 * @param params the completion params that provide the file and cursor position to get the context for
+	 * @param utils the jdt utils
+	 * @param monitor the progress monitor
+	 * @return the cursor context for the given file and cursor position, or null if none of the cursor contexts are applicable
+	 * @throws JavaModelException when the buffer for the file cannot be accessed or the Java model cannot be accessed
+	 */
+	public JavaCursorContext snippetCursorContext(MicroProfileJavaCompletionParams params, IJDTUtils utils,

```

Comment:
Instead of returning the JavaCursorContext (which should be renamed with `JavaCursorContextKind`), I will create a new class JavaCursorContextResult which will host:

 - the kind (ON_CLASS_ANNOTATION, etc)
 - in case of annotation, the offset (or better the position) where the import of annotation type must be done if class doesn't import the annotation type. After that, the snippet completion could return too the additionalTextEdit to insert the proper import.

Reply:
I opened https://github.com/eclipse/lsp4mp/issues/339 based on this discussion. Also, here is a branch with some of my initial work that I removed from this PR: https://github.com/datho7561/lsp4mp/tree/import-offset-for-snippets